src/notebook/widget.ts

```diff
@@ -1223,33 +1226,69 @@ class Notebook extends StaticNotebook {
       target = target.parentElement;
     }
 
-    // Find the target cell and insert the copied cells.
-    let index = this._findCell(target);
-    if (index === -1) {
-      index = this.widgets.length;
-    }
-    let model = this.model;
-    let values = event.mimeData.getData(JUPYTER_CELL_MIME);
-    let factory = model.contentFactory;
-
-    // Insert the copies of the original cells.
-    each(values, (cell: nbformat.ICell) => {
-      let value: ICellModel;
-      switch (cell.cell_type) {
-      case 'code':
-        value = factory.createCodeCell({ cell });
-        break;
-      case 'markdown':
-        value = factory.createMarkdownCell({ cell });
-        break;
-      default:
-        value = factory.createRawCell({ cell });
-        break;
+    let source: Notebook = event.source
+    if(source === this) {
+      // Handle the case where we are moving cells within
+      // the same notebook.
+      event.dropAction = 'move';
+      let toMove: BaseCellWidget[] = event.mimeData.getData('internal:cells');
+
+      //Compute the to/from indices for the move.
+      let fromIndex = ArrayExt.firstIndexOf(this.widgets, toMove[0]);
+      let toIndex = this._findCell(target);
+      if(toIndex === -1) {
+        toIndex = this.widgets.length;
       }
-      model.cells.insert(index++, value);
-    });
-    // Activate the last cell.
-    this.activeCellIndex = index - 1;
+      //Don't move if we are within the block of selected cells.
+      if(toIndex-fromIndex > 0 && toIndex-fromIndex <= toMove.length) {
+        return;
+      }
+      // Move the cells one by one
+      this.model.cells.beginCompoundOperation();
+      if(fromIndex < toIndex) {
+        each(toMove, (cellWidget)=> {
+          this.model.cells.move(fromIndex, toIndex);
+        });
+      } else if (fromIndex > toIndex) {
+        each(toMove, (cellWidget)=> {
+          this.model.cells.move(fromIndex++, toIndex++);
```

Comment:
A subtle gotcha from that link @blink1073: the move semantics are different from those used in phosphor and `ObservableVector`. 

For `CollaborativeList`:
```typescript
let data = [0, 1, 2]
data.move(1,2) // no-op: [0, 1, 2]
data.move(1, 3) // [0, 2, 1]

```

Phosphor:
```typescript
let data = [0, 1, 2]
ArrayExt.move(data, 1, 2) // [0, 2, 1]
```